### PR TITLE
Oracle implementation, tests, commission fees

### PR DESCRIPTION
**Pull request by Holymir**

### CODE REVIEW

File: contracts/Transfer/IPClaimsTransfer.sol [hunk 1]
```diff
@@ -77,15 +112,17 @@ contract IPClaimsTransfer is IIPClaimsTransfer, OwnableUpgradableImplementation
      * @param ipClaimContractAddress - the address of the claim
      * @param signedDataByOwner - hash of the signed data /seller, buyer, IPClaim address, amount in ethers/
      */
-    function transferIPClaimWithEthers(address ipClaimCurrentOwner, address ipClaimContractAddress, bytes signedDataByOwner) public payable onlyValidAddress(ipClaimContractAddress) onlyClaimOwner(ipClaimCurrentOwner, ipClaimContractAddress) {
-        bytes32 bytes32Message = keccak256(abi.encodePacked(ipClaimCurrentOwner, msg.sender, ipClaimContractAddress, msg.value));
+    function transferIPClaimWithEthers(address ipClaimCurrentOwner, address ipClaimContractAddress, uint256 _claimPrice, bytes signedDataByOwner) public payable onlyValidAddress(ipClaimContractAddress) onlyClaimOwner(ipClaimCurrentOwner, ipClaimContractAddress) {
```
Comment (vladichhh) [inline contracts/Transfer/IPClaimsTransfer.sol]: add handling using lastRate

File: contracts/Transfer/IPClaimsTransfer.sol [hunk 2]
```diff
@@ -96,22 +133,27 @@ contract IPClaimsTransfer is IIPClaimsTransfer, OwnableUpgradableImplementation
...
      * @param ipClaimCurrentOwner - current owner of the claim
-     * @param ipClaimContractAddress - the address of the claim
-     * @param amount - amount of the tokens
+     * @param ipClaimContractAddress transferFeeInTokens- the address of the claim
+     * @param _claimPrice - _claimPrice of the tokens
      * @param signedDataByOwner - hash of the signed data /seller, buyer, IPClaim address, amount in tokens/
      */
-    function transferIPClaimWithTokens(address ipClaimCurrentOwner, address ipClaimContractAddress, uint amount, bytes signedDataByOwner) public onlyValidAddress(ipClaimContractAddress) onlyClaimOwner(ipClaimCurrentOwner, ipClaimContractAddress) {
-        bytes32 bytes32Message = keccak256(abi.encodePacked(ipClaimCurrentOwner, msg.sender, ipClaimContractAddress, amount));
+    function transferIPClaimWithTokens(address ipClaimCurrentOwner, address ipClaimContractAddress, uint256 _claimPrice, bytes signedDataByOwner) public onlyValidAddress(ipClaimContractAddress) onlyClaimOwner(ipClaimCurrentOwner, ipClaimContractAddress) {
```
Comment (vladichhh) [inline contracts/Transfer/IPClaimsTransfer.sol]: add handling using lastRate

File: contracts/Transfer/IPClaimsTransfer.sol [hunk 3]
```diff
@@ -96,22 +138,32 @@ contract IPClaimsTransfer is IIPClaimsTransfer, OwnableUpgradableImplementation
...
         address messageSigner = recover(bytes32Message, signedDataByOwner);
         require(messageSigner == ipClaimCurrentOwner);
 
-        uint commissionFee = (amount.mul(commission)).div(100);
+        uint256 transferFeeInTokens = tokenExchangeOracle.convertUSDToTokens(feeInUSD);
+        uint commissionFee = (_claimPrice.mul(commissionPercentage)).div(100);
 
-        assert(ERC20(erc20Instance).transferFrom(msg.sender, ipClaimCurrentOwner, amount.sub(commissionFee))); 
-        assert(ERC20(erc20Instance).transferFrom(msg.sender, feeReceiver, commissionFee));
+        if (!ERC20(erc20Instance).transferFrom(msg.sender, feeReceiver, commissionFee.add(transferFeeInTokens)) ||
```
Comment (thcrnk) [inline contracts/Transfer/IPClaimsTransfer.sol]: Use allowance for the check

File: contracts/Transfer/IPClaimsTransfer.sol (hunk 1 above)
Comment (Holymir) [reply contracts/Transfer/IPClaimsTransfer.sol]: done

**Pull request merged**